client.js

```diff
@@ -0,0 +1,254 @@
+            if (this.channelMessageSubscribers[channelId])
+              this.channelMessageSubscribers[channelId](messagesByChannelId[channelId]);
+          }
+        });
+
+        resolve();
+      };
+
+      const onError = (error) => {
+        reject(error);
+      };
+
+      // Node socket uses different methods than browser socket
+      if (socket.on) {
+        socket.on('connect', onOpen);
+        socket.on('error', onError);
+      } else {
+        socket.onopen = onOpen;
+        socket.onerror = onError;
+      }
+    });
+  }
+
+  disconnect() {
+    if (!this.client) return;
+
+    Object.values(this.client.rpcs).forEach(({ transport }) => {
+      transport.emit('close');
+    });
+
+    delete this.client;
+  }
+
+  isConnected() {
+    return !!this.client;
+  }
+
+  importAuthKeys(authKeys) {
+    this.client.importAuthKeys(authKeys);
+  }
+
+  getAuthKeys() {
+    return this.client.getAuthKeys();
+  }
+
+  subscribeToChannelMessages(channelId, callback) {
+    this.channelMessageSubscribers[channelId] = callback;
+  }
+
+  unsubscribeFromChannelMessages(channelId) {
+    delete this.channelMessageSubscribers[channelId];
+  }
+
+  async request(...args) {
+    return await this.client
+      .call(...args)
+      .catch(async error => {
+        if (error.error_message.includes('_MIGRATE_')) {
+          const [_, nextDcId] = error.error_message.split('_MIGRATE_'); //eslint-disable-line
+          this.client.setDefaultDc(Number(nextDcId));
+          return await this.request(...args);
+        }
+        if (error.error_message === 'AUTH_RESTART') {
+          return await this.request(...args);
+        }
+        if (error.error_message.includes('FLOOD_WAIT')) {
+          const seconds = error.error_message.split('_WAIT_')[1];
+          throw new Error(`Our servers busy talking to Telegram. Please try again in ${seconds} seconds`);
+        }
+        throw new Error(error.error_message);
+      });
+  }
+
+  async checkIfLoggedIn(){
+    return await this.request('account.getAccountTTL', {});
+  }
+
+  async startLogin({mobile}){
+    const { phone_code_hash } = await this.request(
+      'auth.sendCode',
+      {
+        phone_number: mobile,
+        settings: {
+          _: 'codeSettings',
+        }
+      },
+    );
+    return { phoneCodeHash: phone_code_hash };
+  }
+
+  async completeLogin({ mobile, phoneCode, phoneCodeHash }) {
+    const response = await this.request(
+      'auth.signIn',
+      {
+        phone_code: phoneCode,
+        phone_number: mobile,
+        phone_code_hash: phoneCodeHash,
+      },
+    );
+
+    if (response._ === 'auth.authorizationSignUpRequired') {
+      const signUpRequiredMessage = `
+        There is no Telegram account associated with ${mobile}.
+        Please download the Telegram app on your mobile device and sign up, then try again.
+      `;
+      throw new Error(signUpRequiredMessage);
+    }
+    return response;
```

Comment:
why not extract what we need from this response object? its not clear from this code what to expect in the return value of `completeLogin`. Why not store the authkeys I suspect you get back in the `response` object?

Reply:
The response object doesn't have the authKeys, and we don't actually use the response anywhere, I'll just not return anything.